crates/kona-providers/src/blob_provider.rs

```diff
@@ -0,0 +1,150 @@
+#[derive(Debug)]
+pub struct InnerBlobProvider {
+    /// Maximum number of blobs to keep in memory.
+    capacity: usize,
+    /// Order of key insertion for oldest entry eviction.
+    key_order: VecDeque<B256>,
+    /// Maps block hashes to blobs.
+    blocks_to_blob: HashMap<B256, Vec<Blob>>,
+}
+
+impl InnerBlobProvider {
+    /// Creates a new [InnerBlobProvider].
+    pub fn with_capacity(cap: usize) -> Self {
+        Self {
+            capacity: cap,
+            blocks_to_blob: HashMap::with_capacity(cap),
+            key_order: VecDeque::with_capacity(cap),
+        }
+    }
+
+    /// Inserts multiple blobs into the provider.
+    pub fn insert_blobs(&mut self, block_hash: B256, blobs: Vec<Blob>) {
+        if let Some(existing_blobs) = self.blocks_to_blob.get_mut(&block_hash) {
+            existing_blobs.extend(blobs);
+        } else {
+            if self.blocks_to_blob.len() >= self.capacity {
+                if let Some(oldest) = self.key_order.pop_front() {
+                    self.blocks_to_blob.remove(&oldest);
+                }
+            }
+            self.blocks_to_blob.insert(block_hash, blobs);
+        }
+    }
+}
+
+impl LayeredBlobProvider {
+    /// Creates a new [LayeredBlobProvider] with a local blob store, an online primary beacon
+    /// client and an optional fallback blob archiver for fetching blobs.
+    pub fn new(beacon_client_url: Url, blob_archiver_url: Option<Url>) -> Self {
+        let memory = Arc::new(Mutex::new(InnerBlobProvider::with_capacity(512)));
+
+        let online = OnlineBlobProviderBuilder::new()
+            .with_primary(beacon_client_url.to_string())
+            .with_fallback(blob_archiver_url.map(|url| url.to_string()))
+            .build();
+
+        Self { memory, online }
+    }
+
+    /// Inserts multiple blobs into the in-memory provider.
+    pub fn insert_blobs(&mut self, block_hash: B256, blobs: Vec<Blob>) {
+        self.memory.lock().insert_blobs(block_hash, blobs);
+    }
+
+    /// Attempts to fetch blobs using the in-memory blob store.
+    async fn memory_blob_load(
+        &mut self,
+        block_ref: &BlockInfo,
+        hashes: &[IndexedBlobHash],
+    ) -> eyre::Result<Vec<Blob>> {
+        let locked = self.memory.lock();
+
+        let blobs_for_block = locked
+            .blocks_to_blob
+            .get(&block_ref.hash)
+            .ok_or_else(|| eyre!("Blob not found for block ref: {:?}", block_ref))?;
+
+        let mut blobs = Vec::new();
+        for _ in hashes {
+            for blob in blobs_for_block {
+                blobs.push(*blob);
+            }
+        }
+
+        Ok(blobs)
+    }
+
+    /// Attempts to fetch blobs using the online blob provider.
+    #[allow(unused)]
+    async fn online_blob_load(
+        &mut self,
+        block_ref: &BlockInfo,
+        blob_hashes: &[IndexedBlobHash],
+    ) -> Result<Vec<Blob>, BlobProviderError> {
+        self.online.get_blobs(block_ref, blob_hashes).await
```

Comment:
Think we'll want to metrice anytime we end up using the online fallback - but can do that down the road.

Thinking we'll want to use some sort of metrics feature flag like [how I set up in `kona-derive`](https://github.com/ethereum-optimism/kona/blob/main/crates/derive/src/metrics.rs).

Reply:
Yup, I'll make a telemetry issue rn!